### PR TITLE
Add explicit message ack/nack

### PR DESCRIPTION
**Pull request by necrophonic**

### CODE REVIEW

File: app/response_processor.py [hunk 1]
```diff
@@ -1,132 +1,90 @@
...
-            if json:
-                r = session.post(request_url, json=json, headers=headers, verify=verify, auth=auth)
-            else:
-                r = session.get(request_url, headers=headers, verify=verify, auth=auth)
+            self.logger.info("Calling service", request_url=endpoint)
+            res = session.post(endpoint, data=data, headers=headers, verify=False, auth=auth)
 
-            return r
+            if res.status_code == 400:
+                self.logger.error("Receipt rejected by endpoint", request_url=res.url, status_code=400)
```
Comment (eldeal) [inline app/response_processor.py:76]: log tx_id if we have it?

File: app/response_processor.py [hunk 2]
```diff
@@ -1,132 +1,90 @@
...
+                raise BadMessageError("Failure to send receipt")
 
-        except MaxRetryError:
-            self.logger.error("Max retries exceeded (5)", request_url=request_url)
+            elif res.status_code != 200 and res.status_code != 201:
+                # Endpoint may be temporarily down
+                self.logger.error("Bad response from endpoint", request_url=res.url, status_code=res.status_code)
```
Comment (eldeal) [inline app/response_processor.py:81]: log tx_id if we have it

File: app/response_processor.py [hunk 3]
```diff
@@ -1,132 +1,90 @@
...
+                raise RetryableError("Bad response from endpoint")
 
-    def response_ok(self, res):
-        if res.status_code == 200 or res.status_code == 201:
-            self.logger.info("Returned from service", request_url=res.url, status_code=res.status_code)
-            return True
+            else:
+                self.logger.info("Returned from service", request_url=res.url, status_code=res.status_code)
```
Comment (eldeal) [inline app/response_processor.py:85]: log tx_id if we have it

File: tests/test_response_processor.py [hunk 4]
```diff
@@ -1,86 +1,185 @@
...
+                session_mock.return_value = MockResponse(status=500)
+                processor._send_receipt(self.decrypted, self.xml)
+
+    def test_with_400_response(self):
+        with self.assertRaises(BadMessageError):
+            with mock.patch('app.response_processor.session.post') as session_mock:
+                session_mock.return_value = MockResponse(status=400)
+                processor._send_receipt(self.decrypted, self.xml)
+
+# import base64
```
Comment (eldeal) [inline tests/test_response_processor.py]: Why is all of this still here?

**Review by eldeal — CHANGES_REQUESTED** [review]

File: app/response_processor.py (hunk 3 above)
Comment (necrophonic) [reply app/response_processor.py:85]: Should already have it - gets bound to the logger in `_validate` once we know we have it (line 41) [same for others in this package]

**Review by eldeal — APPROVED** [review]

**Pull request merged**